dashboard/modules/runtime_env/runtime_env_agent.py

```diff
@@ -75,63 +78,71 @@ def run_setup_with_logger():
             return await loop.run_in_executor(None, run_setup_with_logger)
 
         serialized_env = request.serialized_runtime_env
-        if serialized_env in self._env_cache:
-            serialized_context = self._env_cache[serialized_env]
-            result = self._env_cache[serialized_env]
-            if result.success:
-                context = result.result
-                logger.info("Runtime env already created successfully. "
-                            f"Env: {serialized_env}, context: {context}")
-                return runtime_env_agent_pb2.CreateRuntimeEnvReply(
-                    status=agent_manager_pb2.AGENT_RPC_STATUS_OK,
-                    serialized_runtime_env_context=context)
-            else:
-                error_message = result.result
-                logger.info("Runtime env already failed. "
-                            f"Env: {serialized_env}, err: {error_message}")
+
+        if serialized_env not in self._env_locks:
+            # async lock to prevent the same env being concurrently installed
+            self._env_locks[serialized_env] = asyncio.Lock()
+
+        async with self._env_locks[serialized_env]:
```

Comment:
The diff is hard to read, but these are the only lines I've added.  The rest of the method is wrapped in the async lock.